src/renderer/api/catalog-entity-registry.ts

```diff
@@ -19,38 +19,59 @@
  * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
  */
 
-import { computed, makeObservable, observable } from "mobx";
+import { computed, observable, makeObservable, action } from "mobx";
 import { subscribeToBroadcast } from "../../common/ipc";
 import { CatalogCategory, catalogCategoryRegistry, CatalogCategoryRegistry, CatalogEntity, CatalogEntityData, CatalogEntityKindData } from "../../common/catalog";
 import "../../common/catalog-entities";
-import { iter } from "../utils";
 import type { Cluster } from "../../main/cluster";
 import { ClusterStore } from "../../common/cluster-store";
 
 export class CatalogEntityRegistry {
-  protected rawItems = observable.array<CatalogEntityData & CatalogEntityKindData>();
-  @observable.ref activeEntity?: CatalogEntity;
+  protected _entities = observable.map<string, CatalogEntity>([], { deep: true });
+  @observable.ref activeEntity: CatalogEntity;
 
   constructor(private categoryRegistry: CatalogCategoryRegistry) {
     makeObservable(this);
   }
 
   init() {
     subscribeToBroadcast("catalog:items", (ev, items: (CatalogEntityData & CatalogEntityKindData)[]) => {
-      this.rawItems.replace(items);
+      this.updateItems(items);
+    });
+  }
+
+  @action updateItems(items: (CatalogEntityData & CatalogEntityKindData)[]) {
+    const newIds = items.map((item) => item.metadata.uid);
+
+    Array.from(this._entities.keys()).forEach((uid) => {
+      if (!newIds.includes(uid)) this._entities.delete(uid);
+    });
+
+    items.forEach((item) => {
+      const existing = this._entities.get(item.metadata.uid);
+
+      if (!existing) {
+        const entity = this.categoryRegistry.getEntityForData(item);
+
+        if (entity) this._entities.set(entity.metadata.uid, entity);
```

Comment:
This breaks the race condition I was talking about earlier. We still want items that were previously added to appear.

Reply:
This would require keeping `rawItems` around

Reply:
Good point, refactored.